### DISCUSSION TITLE
RGCNConv layer with no features

### DISCUSSION THREAD

**Discussion by Laolga** [pyg-team/pytorch_geometric#2449] [Q&A]
Hi!
I'm trying to adjust the graph_saint example from here https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppred/mag/graph_saint.py  such that RGCNConv one-dimensional node index tensor (in which case input features are treated as trainable node embeddings). 
So I did:
```{python}
nodes = data.edge_index_dict[('author', 'writes', 'paper')][1].tolist()
nodes = list(set(nodes))
nodes = torch.tensor(nodes, dtype=torch.long)
data.x_dict["paper"] = nodes
```
And then set in_channels = data.x_dict['paper'].shape[0]
Then when I tried to run the model I got:
RuntimeError: [enforce fail at CPUAllocator.cpp:65] . DefaultCPUAllocator: can't allocate memory: you tried to allocate 3342172169844 bytes. Error code 12 (Cannot allocate memory)

I suspect this is because I did something wrong previously. Could you please explain why it happens?
Also, it's written in the documentation that x can be of type tuple denoting source and destination node features. What exactly does it mean and why one might need that?

Thanks!

**Comment by rusty1s** [comment]
The GraphSAINT example already makes use of trainable embeddings (see `emb_dict`). I'm not really sure how you want to modify it, but I guess you can simple define node indices as 
```python
data.x_dict['paper'] = torch.arange(num_paper_nodes)
```
and pass this vector to the `RGCNConv`.

Separating source and destination node features is useful in a bipartite graph, where nodes are of different type and you want to send messages from one type to another.

**Reply by Laolga** [reply]
Thanks! That's exactly what I wanted, but what is then in_channels? If I set 
```{python}
in_channels = data.x_dict['paper'][0]  #736389
```
I will get 
```{python}
RuntimeError: [enforce fail at CPUAllocator.cpp:65] . DefaultCPUAllocator: can't allocate memory: you tried to allocate 3342172169844 bytes. Error code 12 (Cannot allocate memory
```
Does it make sense to you? In the original example data.x_dict is 736389 by 128 and that should require more memory I guess than just a vector, but it works fine.

**Reply by Laolga** [reply]
Intuitively I guess in_channels should be equal to 1, but in this case I get
```{python
   1688     if input.dim() == 2 and bias is not None:
   1689         # fused op is marginally faster
-> 1690         ret = torch.addmm(bias, input, weight.t())
RuntimeError: mat1 dim 1 must match mat2 dim 0
```

**Reply by rusty1s** [reply]
`in_channels` need to be set to the number of nodes in this case, and this will create a `736389 x 128` weight matrix for *each* relation. Given your number of relations might be high, this can consume a lot of CPU memory. Another way to solve this problem is to utilize an embedding matrix before passing the node features to the `RGCNConv` layer, e.g.:
```python
emb = torch.nn.Embedding(num_nodes, 128)
conv = RGCNConv(128, 128, ...)

x = emb(torch.arange(num_paper_nodes)
x = conv(x, edge_index, edge_type)
```
